src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java

```diff
@@ -0,0 +1,203 @@
+            );
+        } catch (URISyntaxException | StorageException exception) {
+            log.error("Unable to get client for {} container", containerName, exception);
+        }
+    }
+
+    private Optional<BlobInputStream> getBlobInputStream(CloudBlobContainer blobContainer, String zipFileName) {
+        try {
+            return Optional.of(
+                blobContainer
+                    .getBlockBlobReference(zipFileName)
+                    .openInputStream()
+            );
+        } catch (URISyntaxException | StorageException exception) {
+            log.error(
+                "Unable to get blob input stream. Container: {}, blob: {}",
+                blobContainer.getName(),
+                zipFileName
+            );
+
+            return Optional.empty();
+        }
+    }
+
+    private Optional<ZipFileProcessingResult> processInputStream(
+        BlobInputStream blobInputStream,
+        String containerName,
+        String zipFileName
+    ) {
+        try (ZipInputStream zis = new ZipInputStream(blobInputStream)) {
+            return Optional.of(
+                zipFileProcessor.process(zis, containerName, zipFileName)
+            );
+        } catch (DocSignatureFailureException exception) {
+            log.warn(
+                "Rejecting blob - invalid signature. File: {}, Container: {}",
+                zipFileName,
+                containerName,
+                exception
+            );
+
+            handleEventRelatedError(DOC_SIGNATURE_FAILURE, containerName, zipFileName, exception.getMessage());
+
+            blobManager.tryMoveFileToRejectedContainer(zipFileName, containerName, null);
+        } catch (IOException exception) {
+            log.error("Failure reading zip. File: {}, Container: {}", zipFileName, containerName, exception);
+
+            handleEventRelatedError(DOC_FAILURE, containerName, zipFileName, exception.getMessage());
+        }
+
+        return Optional.empty();
+    }
+
+    private boolean uploadParsedZipFileName(Envelope envelope, List<Pdf> pdfs) {
+        try {
+            documentProcessor.uploadPdfFiles(pdfs, envelope.getScannableItems());
+
+            log.info("Uploaded pdfs. File {}, container: {}", envelope.getZipFileName(), envelope.getContainer());
+
+            return true;
+        } catch (Exception exception) {
+            log.error(
+                "Failed to upload PDF files to Document Management. File: {}, Container: {}",
+                envelope.getZipFileName(),
+                envelope.getContainer(),
+                exception
+            );
+
+            envelope.setUploadFailureCount(envelope.getUploadFailureCount() + 1);
+            envelopeRepository.saveAndFlush(envelope);
+            // no need to create event. envelopeProcessor::handleEvent does that
+
+            return false;
+        }
+    }
+
+    private void handleEventRelatedError(Event event, String containerName, String zipFileName, String reason) {
```

Comment:
Could this be done by EnvelopeProcessor instead? It's already used for that. Currently, there's some inconsistence in handling. Whenever we've using EnvelopeProcessor, we don't save reason, but we're doing it here.